napari/utils/naming.py

```diff
@@ -42,3 +46,60 @@ def inc_name_count(name):
         Numbered name incremented by ``1``.
     """
     return numbered_patt.sub(_inc_name_count_sub, name, count=1)
+
+
+def magic_name(value, *, level=1):
+    """Fetch the name of the variable with the given value passed to the calling function.
+
+    Parameters
+    ----------
+    value : any
+        The value of the desired variable.
+    level : int, kwonly, optional
+        The level of nestedness to traverse.
+
+    Returns
+    -------
+    name : str or None
+        Name of the variable, if found.
+    """
+    if level < 1:
+        raise ValueError('cannot have a level lower than 1')
+
+    if not os.getenv('MAGICNAME'):
+        return
+
+    level += 1
+    frame = inspect.currentframe()
+
+    for i in range(level):
+        frame = frame.f_back
```

Comment:
It might be worth trying to make the function recursive (search *each* frame *up to* `level`), so that you don't have to `guess_name` both in `view_image` and `add_image`. This would pick the lowest frame for which the variable is found as long as the name is not `data`. *OR*, it could be the lowest frame that isn't inside the napari library...? Making this function only be called in `AddLayersMixin` would be a massive win for the codebase, and I think the problem is probably solvable...! What do you think?